Migrating keystores from JKS to PKCS12, fixed TCK tests

dmatej opened the pull request:
### PKCS12
* Fixes #25529 , we are using PKCS12 now.
* Some places like AMX had own management of keystores not reflecting changes over years.

### Default security files removed
* Removes "default" files in security.jar
  * Were sometimes used when files were missing like in embedded. cacerts was just an empty keystore. Other files seem unused as tests are passing.
  * I did not like that runtime is parasiting on a set of files without any maintenance, which were sometimes replaced, sometimes not.

### Default timeouts removed
* asadmin start-domain/stop-domain doesn't have default timeout any more. When using autodeploy directory to deploy hundreds war files, we are closing to one minute. 
* It is not easy to guess the good timeout, we always had problems with it, and 7.1 now has the --timeout attribute for that, added in #25580 .
* By default we don't expect that the startup should time out. It should just do what is needed or crash if it cannot.

### Tests
* Updated and synchronized TCK
* Adapted to usage of p12 files as many have the JKS hardcoded.
* All TCK tests now use the same version of Arquillian Containers, 2.0.0 from OmniFish (originally from OmniFaces).
* Changes to Platform TCK were already merged to master branch, however not released yet.
* Changes to Standalone TCKs will be distributed later to their project repositories and jenkins builds, this PR served as a prototype of these changes. After releases of TCKs we can just delete the code from poms (mostly filtering xml and property files).

dmatej committed:
Removed default keystores

- Dangerous and insecure

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Removed obsoleted unused keystore files

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Reviewed create-auth-realm man page

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
SecuritySupport - key stores

- provide an empty keystore if the file does not exist or not set
- log warning if the file is set but does not exist, but continue

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej commented:
Ok, I followed the idea, it broke some pieces, after their resolving embedded should have bit faster starts, however it still broke few tests as now embedded doesn't contain certificates by default. Then tests using HTTPS cannot pass, however they lead to the need to provide an API to enable secured endpoints.

dmatej commented:
> > ```
> > * Changes DN of certificates (Ottawa -> Brussels, Canada -> Belgium)
> > ```
> 
> Q1: Why?

Eclipse Foundation moved to EU few years ago, all active GF developers are in Europe or Japan. However the second question is more interesting.

> Q2: Would the certificate work as good without geo attributes in DN at all?

It seems it would, those attributes are optional. Basically we need just CN which should be a hostname, which is much more complicated question ... so for the "example named domain1" certificate we use "localhost".

I am still working on it, it will not go to 7.0.x, maybe some commits could, but no radical changes.

dmatej committed:
Removed code causing overriding keystores in embedded

- For embedded, keystores must be always provided by the user (or we can run
  without TLS support etc)

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
SystemTasksImpl - don't set DAS defaults to embedded

- note: even version is not available for embedded - we should improve it.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

hs536 reviewed (commented):
My apologies, the above approval was a misoperation.

dmatej commented:
> My apologies, the above approval was a misoperation.

No problem, I know it is not ready to be reviewed+merged, I will probably split it later to smaller PRs.

dmatej committed:
SecureAdminConfigUpgrade - cleanup, using KeyTool

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Removed unused BAD_DELETE_TEMP_CERT_FILE logging constant

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Removed two unused methods used for nonexisting templates

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Using same constants for default keystore password and type

- note that type can be autodetected both by keytool and KeyStore

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Workaround for issue mentioned in #25529 - PKCS12/JKS vs private key passwords

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Using constants for default keystore type, file names, password

- in few tests changed to different to avoid confusion

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej commented:
Status: Time to stop now.
* domain1 contains keystores generated in build. User should create new domain for production usage.
* embedded doesn't contain anything, BUT it is possible to set own keystore+truststore
* default keystore type is PKCS12, not JKS
* The keystore type can be detected from the file name (tested P12, PFX, JKS, JCEKS - last two are legacy formats, so we avoid them, but it is possible).
* new class KeyTool wraps keytool program; most tasks can be done in Java already; however P12 and JKS formats have bit different behavior when changing passowrds. See #25529 for more. KeyTool was ideal to simplify that, so we don't have to care now.
* Reduced copy and paste - using constants instead of hardcoded strings and copy pasted constants. However SystemPropertyConstants class should die, now it resolves time consuming "search and replace" problem.
* master-password, domains-passwords - use .p12 file extension now (used JCEKS, in Java5 was the only impl supporting keys and their CRUD).

TODO: 
* Not sure about the TCK, I have to try. 
* Now it is 47 commits - I will split at least part to own PRs.

dmatej committed:
Fixed situaion when keystores don't exist

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
KeyTool - implemented copying certificate to another keystore + usages

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
keyTool.changeKeystorePassword and changeKeyPassword

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej commented:
Status: seems all tests including TCKs on Eclipse and my infrastructure are passing.

dmatej committed:
New KeyTool adapter

- no dependencies on other GF sources
- output is not read in parallel threads
- we don't care about the language of the system, we just write password 4 times
- then we read the output, basically we ignore it except logs.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
More KeyTool class usages

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Removed default admin-keyfile and keyfile

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Removed default login.conf

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej commented:
Status: passed TCKs using integration in the project using my hardware; passed also updated platform TCK on [Eclipse Jenkins](https://ci.eclipse.org/jakartaee-tck/job/jakartaee-tck/job/PR-2374/), however standalone TCKs were not touched yet (our integration does it).

dmatej committed:
Removed default policy files

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Removed nonexisting keystore.jks and cacerts.jks

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Minor cleanups

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Fixed tests depending on s1as certificate and key

- basically tests working with HTTP over TLS

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Reducing complexity and duplications around keystores

- types and passwords are not respected everywhere (still)
- we need to have control over this
- also deleted some unused code, especially constants

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Switching from JKS and JCEKS to PKCS12

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
TCK - changed groupId of the arquillian plugin, updated

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
TCK cleanup

- Option to upload tck-downloads to a local Nexus cache
  - helps to save dependency on network and downloading huge packages
  - no need to run this part again on rebuild
  - option to push that to a separate jenkins job
- When using deploy plugin, do the upload in the end
- Workaround for maven-war-plugin and artifacts without web.xml and servlets
  - the plugin uses servlet dependency to guess the spec, however now the
    servlet is not a mandatory dependency for a war file.
- Fixed warning about property file encoding
  - usually we use Latin1 as in old JDKs, but using UTF-8 shouldn't harm here
- Synchronzied dependencies
  - Updated shrinkwrap to 3.3.4 and its "relatives" which were sometimes
    overriden by other dependency
  - Arquillian 2.0.0
- Removed connectors-full tck - already covered by TckRunner
- Updated Tags TCK
- Synchronized TCK standalone projects to use same rules, common dependencies
  and same macros.
- Faces TCK tricked to follow our dependency tree
- Arquillian extension changed group id

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

# Conflicts:
#	nucleus/parent/pom.xml

dmatej commented:
And yet once more to draft again - when I was rebasing this I forgot two commits with the TCK changes, meanwhile I was on the long journey with race conditions and arquillian - which I have to release before merging this.

dmatej committed:
Resolving keytool timeouts on CI

- 60 seconds instead of 10
- Printing output in error messages

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Removed default timeout for start and stop domain

- By default we are sure what we are doing
- It is not easy to estimate how much time is needed.
- Can be explicitly set using --timeout argument

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Using released 2.0.0 OmniFish Arquillian artifacts in TCK

- Fixed also CDI TCK using wrong package

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej commented:
Status: Done, I am running TCKs on private Jenkins for the last time. Two days ago few blocks failed: CDI (forgot to rename class), EJB (Jenkins node network issue) and Faces (removed one  of Arquillian GlassFish jars which was required so I had to revert it back), now all of them passed alone, so I am verifying this PR again as a whole. 
The ARQ GF project 2.0.0 was released today to Maven Central.

hs536 commented on nucleus/admin/cli/src/main/java/com/sun/enterprise/admin/cli/CLIConstants.java:
```diff
@@ -35,7 +35,7 @@ public final class CLIConstants {
     public static final String DEFAULT_HOSTNAME = "localhost";
     public static final String EOL = System.lineSeparator();
 
-    public static final Duration WAIT_FOR_DAS_TIME_MS = getEnv(TIMEOUT_START_SERVER, 60L);
```
isn't there a risk of the start commands waiting indefinitely?

avpinchuk reviewed (approved):
LGTM
